File: src/product/Abbot.Web/assets/react/models/schedule.ts
export enum ScheduleType {
    Hourly = 'hourly',
    Daily = 'daily',
    Weekly = 'weekly',
    Monthly = 'monthly',
    Advanced = 'advanced',
}

export type DayOfMonth = number;
export const allDaysOfMonth: DayOfMonth[] = Array.from(Array(31).keys())
    .map(i => i + 1 as DayOfMonth)

export enum Weekday {
    Sunday = 'sunday',
    Monday = 'monday',
    Tuesday = 'tuesday',
    Wednesday = 'wednesday',
    Thursday = 'thursday',
    Friday = 'friday',
    Saturday = 'saturday',
}

export type SerializedSchedule =
    { type: ScheduleType.Hourly, minute: number } |
    { type: ScheduleType.Daily, hour: number, minute: number } |
    { type: ScheduleType.Weekly, hour: number, minute: number, weekdays: Weekday[] } |
    { type: ScheduleType.Monthly, hour: number, minute: number, dayOfMonth: DayOfMonth } |
    { type: ScheduleType.Advanced, cron: string };

export function describeDayOfMonth(value: DayOfMonth) {
    switch (value) {
        case 1: return '1st';
        case 2: return '2nd';
        case 3: return '3rd';
        case 21: return '21st';
        case 22: return '22nd';
        case 23: return '23rd';
        case 31: return '31st';
        default: return `${value}th`;
    }
}

export abstract class Schedule {
    constructor(public type: ScheduleType) {
    }

    static parse(input: SerializedSchedule | string) {
        let serialized: SerializedSchedule;
        if (!input) {
            // Default to daily at noon
            return new DailySchedule(12, 0);
        } else if (typeof input === 'string') {
            // We have to hack things a bit.
            // Any old schedules will be a cron string, so we check to see if this is valid JSON
            try {
                serialized = JSON.parse(input) as SerializedSchedule;
            } catch {
                // Invalid JSON, just use the value as a cron string
                return new AdvancedSchedule(input);
            }
        } else if (typeof input === 'object' && 'type' in input) {
            serialized = input as SerializedSchedule;
        } else {
            throw new Error('Invalid schedule');
        }

        switch (serialized.type) {
            case ScheduleType.Hourly:
                return new HourlySchedule(serialized.minute || 0);
            case ScheduleType.Daily:
                return new DailySchedule(serialized.hour || 0, serialized.minute || 0);
            case ScheduleType.Weekly:
                return new WeeklySchedule(serialized.hour || 0, serialized.minute || 0, serialized.weekdays || []);
            case ScheduleType.Monthly:
                return new MonthlySchedule(serialized.hour || 0, serialized.minute || 0, serialized.dayOfMonth || 1);
            case ScheduleType.Advanced:
                return new AdvancedSchedule(serialized.cron || '10 * * * *');
        }
    }

    static create(type: ScheduleType) {
        switch (type) {
            case ScheduleType.Hourly:
                return new HourlySchedule(0);
            case ScheduleType.Daily:
                return new DailySchedule(0, 0);
            case ScheduleType.Weekly:
                return new WeeklySchedule(0, 0, []);
            case ScheduleType.Monthly:
                return new MonthlySchedule(0, 0, 1);
            case ScheduleType.Advanced:
                return new AdvancedSchedule('10 * * * *');
        }
    }

    abstract get description(): string;
    abstract serialize(): SerializedSchedule;
}

export class HourlySchedule extends Schedule {
    constructor(public minute: number) {
        super(ScheduleType.Hourly);
    }

    serialize(): SerializedSchedule {
        return { type: ScheduleType.Hourly, minute: this.minute };
    }

    get description(): string {
        const mins = this.minute.toLocaleString(undefined, { minimumIntegerDigits: 2 });
        return `Every hour at :${mins}`;
    }
}

export class DailySchedule extends Schedule {
    constructor(public hour: number, public minute: number) {
        super(ScheduleType.Daily);
    }

    serialize(): SerializedSchedule {
        return { type: ScheduleType.Daily, hour: this.hour, minute: this.minute };
    }

    get description(): string {
        const hr = this.hour.toLocaleString(undefined, { minimumIntegerDigits: 2 });
        const mins = this.minute.toLocaleString(undefined, { minimumIntegerDigits: 2 });
        return `Every day at ${hr}:${mins}`;
    }
}

export class WeeklySchedule extends Schedule {
    constructor(public hour: number, public minute: number, public weekdays: Weekday[]) {
        super(ScheduleType.Weekly);
    }

    serialize(): SerializedSchedule {
        return { type: ScheduleType.Weekly, hour: this.hour, minute: this.minute, weekdays: this.weekdays };
    }

    get description(): string {
        const weekdays = this.weekdays.map(w => w[0].toUpperCase() + w.slice(1)).join(', ');
        const hr = this.hour.toLocaleString(undefined, { minimumIntegerDigits: 2 });
        const mins = this.minute.toLocaleString(undefined, { minimumIntegerDigits: 2 });
        return `Every ${weekdays} at ${hr}:${mins}`;
    }
}

export class MonthlySchedule extends Schedule {
    constructor(public hour: number, public minute: number, public dayOfMonth: DayOfMonth) {
        super(ScheduleType.Monthly);
    }

    serialize(): SerializedSchedule {
        return { type: ScheduleType.Monthly, hour: this.hour, minute: this.minute, dayOfMonth: this.dayOfMonth };
    }

    get description(): string {
        const hr = this.hour.toLocaleString(undefined, { minimumIntegerDigits: 2 });
        const mins = this.minute.toLocaleString(undefined, { minimumIntegerDigits: 2 });
        return `Every month on the ${describeDayOfMonth(this.dayOfMonth)} at ${hr}:${mins}`;
    }
}

export class AdvancedSchedule extends Schedule {
    constructor(public cron: string) {
        super(ScheduleType.Advanced);
    }

    serialize(): SerializedSchedule {
        return { type: ScheduleType.Advanced, cron: this.cron };
    }

    get description(): string {
        return `Advanced: ${this.cron}`;
    }
}

File: src/product/Abbot.Web.Library/Playbooks/Schedule.cs
using System.Collections.Generic;
using System.Diagnostics;
using System.Linq;
using Newtonsoft.Json;

namespace Serious.Abbot.Playbooks;

public enum Weekday
{
    Sunday = 0,
    Monday = 1,
    Tuesday = 2,
    Wednesday = 3,
    Thursday = 4,
    Friday = 5,
    Saturday = 6,
}

[JsonConverter(typeof(ScheduleJsonConverter))]
public abstract record Schedule
{
    public abstract string ToCronString();
}

public record HourlySchedule(int Minute) : Schedule
{
    public override string ToCronString() => $"{Minute} * * * *";
}

public record DailySchedule(int Hour, int Minute) : Schedule
{
    public override string ToCronString() => $"{Minute} {Hour} * * *";
}

public record WeeklySchedule(int Hour, int Minute, IReadOnlyList<Weekday> Weekdays) : Schedule
{
    public override string ToCronString() => $"{Minute} {Hour} * * {string.Join(",", Weekdays.Select(w => (int)w))}";
}

public record MonthlySchedule(int Hour, int Minute, DayOfMonth DayOfMonth) : Schedule
{
    public override string ToCronString() => $"{Minute} {Hour} {DayOfMonth.Value} * *";
}

public record AdvancedSchedule(string Cron) : Schedule
{
    public override string ToCronString() => Cron;
}

// TODO: At some point, we'll want to support special days like "last day of the month" and "last weekday of the month".
[JsonConverter(typeof(DayOfMonthConverter))]
public record DayOfMonth(int Value);

public class DayOfMonthConverter : JsonConverter<DayOfMonth>
{
    public override void WriteJson(JsonWriter writer, DayOfMonth? value, JsonSerializer serializer)
    {
        writer.WriteValue(value?.Value);
    }

    public override DayOfMonth? ReadJson(JsonReader reader, Type objectType, DayOfMonth? existingValue,
        bool hasExistingValue,
        JsonSerializer serializer)
    {
        if (reader.TokenType is not JsonToken.Integer || reader.Value is not long l)
        {
            throw new InvalidOperationException(
                $"Could not read DayOfMonth from {nameof(JsonToken)}.{reader.TokenType}.");
        }

        return new((int)l);
    }
}

public class ScheduleJsonConverter : JsonConverter<Schedule>
{
    public override void WriteJson(JsonWriter writer, Schedule? value, JsonSerializer serializer)
    {
        if (value == null)
        {
            writer.WriteNull();
            return;
        }

        var serialized = new SerializedSchedule(value);
        serializer.Serialize(writer, serialized);
    }

    public override Schedule? ReadJson(JsonReader reader, Type objectType, Schedule? existingValue,
        bool hasExistingValue,
        JsonSerializer serializer)
    {
        var serialized = serializer.Deserialize<SerializedSchedule>(reader);

        if (serialized is null)
        {
            return null;
        }

        switch (serialized.Type?.ToLowerInvariant())
        {
            case "hourly":
                return new HourlySchedule(
                    serialized.Minute
                    ?? throw new JsonException($"Missing required property '{nameof(serialized.Minute)}'"));
            case "daily":
                return new DailySchedule(
                    serialized.Hour
                    ?? throw new JsonException($"Missing required property '{nameof(serialized.Hour)}'"),
                    serialized.Minute
                    ?? throw new JsonException($"Missing required property '{nameof(serialized.Minute)}'"));
            case "weekly":
                return new WeeklySchedule(
                    serialized.Hour
                    ?? throw new JsonException($"Missing required property '{nameof(serialized.Hour)}'"),
                    serialized.Minute
                    ?? throw new JsonException($"Missing required property '{nameof(serialized.Minute)}'"),
                    serialized.Weekdays
                    ?? throw new JsonException($"Missing required property '{nameof(serialized.Weekdays)}'"));
            case "monthly":
                return new MonthlySchedule(
                    serialized.Hour
                    ?? throw new JsonException($"Missing required property '{nameof(serialized.Hour)}'"),
                    serialized.Minute
                    ?? throw new JsonException($"Missing required property '{nameof(serialized.Minute)}'"),
                    serialized.DayOfMonth
                    ?? throw new JsonException($"Missing required property '{nameof(serialized.DayOfMonth)}'"));
            case "advanced":
                return new AdvancedSchedule(serialized.Cron
                                            ?? throw new JsonException(
                                                $"Missing required property '{nameof(serialized.Cron)}'"));
            default:
                throw new JsonException($"Unknown schedule type '{serialized.Type}'");
        }
    }

    class SerializedSchedule
    {
        [JsonProperty("type")]
        public string? Type { get; set; }

        [JsonProperty("hour")]
        public int? Hour { get; set; }

        [JsonProperty("minute")]
        public int? Minute { get; set; }

        [JsonProperty("weekdays")]
        public IReadOnlyList<Weekday>? Weekdays { get; set; }

        [JsonProperty("dayOfMonth")]
        public DayOfMonth? DayOfMonth { get; set; }

        [JsonProperty("cron")]
        public string? Cron { get; set; }

        [JsonConstructor]
        public SerializedSchedule()
        {
        }

        public SerializedSchedule(Schedule schedule)
        {
            switch (schedule)
            {
                case HourlySchedule h:
                    Type = "hourly";
                    Minute = h.Minute;
                    break;
                case DailySchedule d:
                    Type = "daily";
                    Hour = d.Hour;
                    Minute = d.Minute;
                    break;
                case WeeklySchedule w:
                    Type = "weekly";
                    Hour = w.Hour;
                    Minute = w.Minute;
                    Weekdays = w.Weekdays;
                    break;
                case MonthlySchedule m:
                    Type = "monthly";
                    Hour = m.Hour;
                    Minute = m.Minute;
                    DayOfMonth = m.DayOfMonth;
                    break;
                case AdvancedSchedule a:
                    Type = "advanced";
                    Cron = a.Cron;
                    break;
                default:
                    throw new UnreachableException($"Unknown schedule type: {schedule.GetType()}");
            }
        }
    }
}
